### ISSUE TITLE
A connection is not put back to the pool if DB.Query panics

### ISSUE THREAD

**Issue by deni64k** [go-pg/pg#137]
Hi guys,
We struggled with a very mystical problem of no available connections in the pool.
After a very long investigation, I figured out that if the first parameter `model` in `DB.Query` is given as an value, `simpleQueryData` will panic. In that case a connection won't be freed.
Though it wasn't a correct way of using, `DB.Query` should always put back a connection to the pool, otherwise the pool will not have an available connection.

Basically,

```
db.freeConn(cn, err)
```

should be

```
defer db.freeConn(cn, err)
```

https://github.com/go-pg/pg/blob/v4/db.go#L151

And I believe, this is not the only place with such problem in the package.

It reminded me of exception safety in C++: https://en.wikipedia.org/wiki/Exception_safety
You'd better use RAII technique (Go's `defer`) to free resources instead of relying on a perfect case.

**Comment by vmihailenco** [comment]
Hey,

I assume that you use `recover` to continue execution, correct?

One possible issue with `defer db.freeConn(cn, err)` is that on panic it will put connection back to the pool, which can cause some side effects if data was not fully written/read.

An alternative would be to fix go-pg to never panic :) Can you please attach a backtrace showing the panic? Thanks.

**Comment by neglectedvalue** [comment]
Yes, we use `gin.Recover` for that. Every `panic` is caught by that and the web application continues working.

In what cases could data be not fully written or read?

This is our backtrace:
```
ERROR: reflect: call of reflect.Value.Type on zero Value
goroutine 2973 [running]:
runtime/debug.Stack(0x0, 0x0, 0x0)
        /usr/local/Cellar/go/1.6/libexec/src/runtime/debug/stack.go:24 +0x80
github.com/housinganywhere/ha/vendor/github.com/housinganywhere/gomvc.GetHandler.func1.1(0x7fd60d47b458, 0xc821757b80)
        /Users/negval/Code/go/src/github.com/housinganywhere/ha/vendor/github.com/housinganywhere/gomvc/gomvc.go:103 +0x284
panic(0xf09200, 0xc821ed9260)
        /usr/local/Cellar/go/1.6/libexec/src/runtime/panic.go:426 +0x4e9
reflect.Value.Type(0x0, 0x0, 0x0, 0x0, 0x0)
        /usr/local/Cellar/go/1.6/libexec/src/reflect/value.go:1671 +0x7b
github.com/housinganywhere/pg.NewColumnLoader(0xfd2120, 0x0, 0x0, 0x0, 0x0, 0x0)
        /Users/negval/Code/go/src/github.com/housinganywhere/pg/loader.go:176 +0x2a8
github.com/housinganywhere/pg.readDataRow(0xc8222974a0, 0xfd2120, 0x0, 0xc822722ae0, 0x6, 0x6, 0x0, 0x0)
        /Users/negval/Code/go/src/github.com/housinganywhere/pg/messages.go:407 +0xe0
github.com/housinganywhere/pg.readSimpleQueryData(0xc8222974a0, 0x7fd60d275860, 0xc8220165b0, 0x0, 0x0, 0x0)
        /Users/negval/Code/go/src/github.com/housinganywhere/pg/messages.go:454 +0x1ee
github.com/housinganywhere/pg.simpleQueryData(0xc8222974a0, 0x7fd60d275860, 0xc8220165b0, 0xc821af9c80, 0x3d, 0xc822516388, 0x2, 0x2, 0x0, 0x0, ...)
        /Users/negval/Code/go/src/github.com/housinganywhere/pg/db.go:413 +0x122
github.com/housinganywhere/pg.(*DB).Query(0xc8201af0f0, 0x7fd60d275860, 0xc8220165b0, 0xc821af9c80, 0x3d, 0xc822516388, 0x2, 0x2, 0x0, 0x0, ...)
        /Users/negval/Code/go/src/github.com/housinganywhere/pg/db.go:233 +0xee
github.com/housinganywhere/pg.(*DB).QueryOne(0xc8201af0f0, 0xfd2120, 0x0, 0xc821af9c80, 0x3d, 0xc822516388, 0x2, 0x2, 0x199, 0x0, ...)
        /Users/negval/Code/go/src/github.com/housinganywhere/pg/db.go:246 +0xec
github.com/housinganywhere/repo-pg.SelectOne(0xfd2120, 0x0, 0xc821af9c80, 0x3d, 0xc822516388, 0x2, 0x2, 0x0, 0x0)
        /Users/negval/Code/go/src/github.com/housinganywhere/repo-pg/repo.go:117 +0x620
...cut...
```

I'd suggest this quick fix:
```		func() {
			res, err = simpleQuery(cn, q, args...)
			defer db.freeConn(cn, err)
		}()
```
Since `defer` calls are executed after function returns, it is better to narrow its scope and put a connection back straight away after using, not when `DB.Query` returns.
